crates/rpc/rpc/src/eth/api/mod.rs

```diff
@@ -448,7 +464,162 @@ struct EthApiInner<Provider, Pool, Network> {
     pending_block: Mutex<Option<PendingBlock>>,
     /// A pool dedicated to blocking tasks.
     blocking_task_pool: BlockingTaskPool,
+    /// Cache for block fees history
+    fee_history_cache: FeeHistoryCache,
     /// An http client for communicating with sequencers.
     #[cfg(feature = "optimism")]
     http_client: reqwest::Client,
 }
+
+/// Settings for the [EthStateCache](crate::eth::cache::EthStateCache).
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FeeHistoryCacheConfig {
+    /// Max number of blocks in cache.
+    ///
+    /// Default is 1024.
+    pub max_blocks: u64,
+}
+
+impl Default for FeeHistoryCacheConfig {
+    fn default() -> Self {
+        FeeHistoryCacheConfig { max_blocks: 1024 }
+    }
+}
+
+/// Wrapper struct for BTreeMap
+#[derive(Debug, Clone)]
+pub struct FeeHistoryCache {
+    lower_bound: Arc<AtomicU64>,
+    upper_bound: Arc<AtomicU64>,
+    config: FeeHistoryCacheConfig,
+    entries: Arc<tokio::sync::RwLock<BTreeMap<u64, FeeHistoryEntry>>>,
+}
+
+impl FeeHistoryCache {
+    /// Creates new FeeHistoryCache instance, initialize it with the mose recent data, set bounds
+    pub fn new(config: FeeHistoryCacheConfig) -> Self {
+        let init_tree_map = BTreeMap::new();
+
+        let entries = Arc::new(tokio::sync::RwLock::new(init_tree_map));
+
+        let upper_bound = Arc::new(AtomicU64::new(0));
+        let lower_bound = Arc::new(AtomicU64::new(0));
+
+        FeeHistoryCache { config, entries, upper_bound, lower_bound }
+    }
+
+    /// Processing of the arriving blocks
+    pub async fn on_new_block<'a, I>(&self, headers: I)
+    where
+        I: Iterator<Item = &'a SealedHeader>,
+    {
```

Comment:
this should take new block because we want to calc the tx percentiles here as well